joblib/numpy_pickle_compat.py

```diff
@@ -0,0 +1,234 @@
+
+    def read(self, unpickler):
+        """Reconstruct the array from the meta-information and the z-file."""
+        # Here we a simply reproducing the unpickling mechanism for numpy
+        # arrays
+        filename = os.path.join(unpickler._dirname, self.filename)
+        array = unpickler.np.core.multiarray._reconstruct(*self.init_args)
+        with open(filename, 'rb') as f:
+            data = read_zfile(f)
+        state = self.state + (data,)
+        array.__setstate__(state)
+        return array
+
+
+class ZipNumpyUnpickler(Unpickler):
+    """A subclass of the Unpickler to unpickle our numpy pickles."""
+
+    dispatch = Unpickler.dispatch.copy()
+
+    def __init__(self, filename, file_handle, mmap_mode=None):
+        """Constructor."""
+        self._filename = os.path.basename(filename)
+        self._dirname = os.path.dirname(filename)
+        self.mmap_mode = mmap_mode
+        self.file_handle = self._open_pickle(file_handle)
+        Unpickler.__init__(self, self.file_handle)
+        try:
+            import numpy as np
+        except ImportError:
+            np = None
+        self.np = np
+
+    def _open_pickle(self, file_handle):
+        return BytesIO(read_zfile(file_handle))
+
+    def load_build(self):
+        """Set the state of a newly created object.
+
+        We capture it to replace our place-holder objects,
+        NDArrayWrapper, by the array we are interested in. We
+        replace them directly in the stack of pickler.
+        """
+        Unpickler.load_build(self)
+        if isinstance(self.stack[-1], NDArrayWrapper):
+            if self.np is None:
+                raise ImportError("Trying to unpickle an ndarray, "
+                                  "but numpy didn't import correctly")
+            nd_array_wrapper = self.stack.pop()
+            array = nd_array_wrapper.read(self)
+            self.stack.append(array)
+
+    # Be careful to register our new method.
+    if PY3:
+        dispatch[pickle.BUILD[0]] = load_build
+    else:
+        dispatch[pickle.BUILD] = load_build
+
+
+def load_compatibility(filename):
+    """Reconstruct a Python object from a file persisted with joblib.dump.
+
+    This function ensures the compatibility with joblib old persistence format
+    (<= 0.9.3).
+
+    Parameters
+    -----------
+    filename: string
+        The name of the file from which to load the object
+
+    Returns
+    -------
+    result: any Python object
+        The object stored in the file.
+
+    See Also
+    --------
+    joblib.dump : function to save an object
+
+    Notes
+    -----
+
+    This function can load numpy array files saved separately during the
+    dump.
+    """
+    with open(filename, 'rb') as file_handle:
+        # We are careful to open the file handle early and keep it open to
+        # avoid race-conditions on renames. That said, if data are stored in
+        # companion files, moving the directory will create a race when
```

Comment:
Same comment as above about companion files, the simplest is to remove the second sentence.